DiscordBM for SSWG Incubation Proccess (#81)
swift-server/sswg: proposals/NNNN-discordbm.md

## Introduction

DiscordBM is a new multi-platform Server-Side focused Swift library for interacting with the Discord API, for making bots.

It's already being used in Vapor's [Penny](https://github.com/vapor/penny-bot) bot for the past 11+ months, and it's been serving the community very nicely.

## Motivation

The Server-Side Swift community lacks a maintained and robust package to communicate with Discord.
This is important because Discord is gaining more and more popularity amongst communities and has already replaced Slack as the go-to place for Open Source communities. A lot of successful projects and languages such as `Rust`, `Python`, `Vue` have chosen Discord as the place for their community.

## Proposed solution

> DiscordBM provides an elegant, robust, type-safe, swifty and [insert positive word here] way to communicate with Discord's bot APIs. This can be useful to bridge the gap between the Swift community and Discord bots.

Reviewer: ```suggestion
DiscordBM provides an elegant, robust, type-safe way to communicate with Discord's bot APIs. Therefore bridging the gap between the Swift community and Discord bots.
```

## Design

> DiscordBM not only comes with the core and necessary means to communicate with Discord, but also has a bunch of convenience or useful stuff that users might otherwise need to implement themselves. I'll try to explain the core implementation in detail, and briefly mention some of the rest.

Reviewer: ```suggestion
DiscordBM not only comes with the core and necessary means to communicate with Discord, but also has a bunch of convenience or useful stuff that users might otherwise need to implement themselves. This document first  explains the core implementation in detail, and then briefly mention some of the surrounding features.
```

```

In the other case where Discord accepts the identification, Discord will start sending the Discord events to you. For more context, there are 70+ types of events and they include a created/updated/deleted message, a create/updated/deleted channel, and many more. DiscordBM allows user access to these events using an `AsyncStream<Gateway.Event>` which users can acquire and use like so:

```swift
...

/// Get an `AsyncStream` of `Gateway.Event`s
let stream = await bot.makeEventsStream()

/// Handle each event in the stream
/// When used in the @main entry point, this stream will never end,
/// therefore preventing your executable from exiting
for await event in stream {
    switch event.data {
    case let .messageCreate(message):
        print("NEW MESSAGE!", message)

        /// You can also use the `GatewayEventHandler` protocol for more convenience
    default: break
    }
}
```

> This is also in line with the latest structured concurrency advancements, and users won't need to block a thread or call functions such as `RunLoop.main.run()`. Generally speaking, DiscordBM has gone full-in on structured concurrency and makes extensive use of actors. The whole library is already in full compliance with `-strict-concurrency=complete`.

Reviewer: Note: Including task cancellation?

Author: For DiscordClient stuff, it's just whatever AsyncHTTPClient does. For gateway, no. I did take a look at kind of rewriting the Gateway logic with Task cancellation support a few months ago, but it didn't go well. Can't exactly remember why, i think i couldn't come up with a plan and came to the conclusion that Task cancellation _might_ not fit.

`GatewayManager` also needs to keep the Discord connection alive. This happens in two ways.

1. The usual websocket ping-pongs.
2. By reconnecting when Discord closes a Websocket connection.

> Connection closures are normal and initiated by Discord when they see fit. The closures could happen 10+ times a day, but are mostly not distruptive. Upon a conneciton closure, DiscordBM will correctly resume the last connection, and Discord will send the events lost when there was no connection.

Reviewer: ```suggestion
Connection closures are normal and initiated by Discord when they see fit. The closures could happen 10+ times a day, but are mostly not distruptive. Upon a connection closure, DiscordBM will correctly resume the last connection, and Discord will send the events that were missed when there was no connection.
```

        channelId: message.channel_id,
        payload: .init(content: "Got a message: '\(message.content)'")
    )
```

And you can safely and effortlessly decode Discord's response to the correct type:

```swift
/// `message` will be of type `DiscordChannel.Message`
let message = try response.decode()
```

The REST API endpoints are all declared as extensions to the `DiscordClient` protocol like so:

```swift
/// https://discord.com/developers/docs/resources/channel#create-message
@inlinable
func createMessage(
    channelId: ChannelSnowflake,
    payload: Payloads.CreateMessage
) async throws -> DiscordClientResponse<DiscordChannel.Message> {
    let endpoint = APIEndpoint.createMessage(channelId: channelId)
    return try await self.sendMultipart(
        request: .init(to: endpoint),
        payload: payload
    )
}
```

This function returns a `DiscordClientResponse`. The generic `C` type that `DiscordClientResponse` takes makes it possible to decode Discord's responses to a library-defined type as you saw in `let message = try response.decode()`:

```swift
/// Represents a Discord HTTP response for endpoints that return some data in the body.

> public struct DiscordClientResponse<C>: Sendable where C: Codable {

Reviewer: Doesn't C also need to be Sendable?

Author: Not really. It would be fine for the library to add that constraint, but this hasn't been necessary.

What happens is when you have a value of type `DiscordClientResponse<C>`, you can get a `C` value with the `.decode()` function, and what `.decode()` function returns is always Sendable enforced by e.g. the `DiscordClient.send()` functions which return the `DiscordClientResponse<C>`.

I'm not against adding that, but it won't have too much of a use-case. The decoded value doesn't really _need_ to be Sendable as far as a decoded-object is concerned.

    /// The raw http response.
    public let httpResponse: DiscordHTTPResponse

    ...

    /// Decodes the response to `C`.
    @inlinable
    public func decode() throws -> C {
        try httpResponse.decode(as: C.self)
    }
}
```

The `DiscordHTTPResponse` type contains a HTTP response:

```swift
/// Represents a raw Discord HTTP response.
public struct DiscordHTTPResponse: Sendable, CustomStringConvertible {

    ...

    public var host: String
    public var status: HTTPResponseStatus
    public var version: HTTPVersion
    public var headers: HTTPHeaders
    public var body: ByteBuffer?

    ...
}
```

This is the basics of `DiscordClient`, but `DefaultDiscordClient` comes with quite a few useful configurations and behaviors as I'll explain next:

```swift
public struct DefaultDiscordClient: Sendable, DiscordClient {
    ...

    let client: HTTPClient
    let configuration: ClientConfiguration
    let cache: ClientCache

    ...
}
```

> - The `cache` is a storage for the cached responses.

Reviewer: What's being cached and why?

Author: Should i explain that more in the proposal?
Caching is disabled by default. It just caches the http responses.
As to why, it's just the same reason you would guess. To do less http requests. It's faster, doesn't consume http limit etc...

>   The `configuration` has options to customize how caching is done.

Reviewer: ```suggestion
  The `configuration` has options, such as customizing how caching or retrying is done.
```

  `ClientCache` is an actor, and it's shared across all `DefaultDiscordClient`s that have the same authorization token.

The configuration options:

```swift
public struct ClientConfiguration: Sendable {
    ...

    public let cachingBehavior: CachingBehavior
    public var retryPolicy: RetryPolicy?
    public var performValidations: Bool

    ...
}
```

- The `cachingBehavior` can be customized for each endpoint with a different time-to-live.
  Whenever there is a successful response, the discord-client will cache the response if allowed by this.
- `performValidations` specifies whether or not to perform client-side validations.
  This is what `ValidatablePayload` protocol is useful for, which adds a `validate()` func to payloads.
  The validations only cover what is mentioned in Discord docs. They don't guarantee a valid payload.
- And there is the `retryPolicy`, which specifies how to retry requests.

>   DiscordBM has some magically-nice behavior built around `retryPolicy`, as I'll explain next.

Reviewer: ```suggestion
  DiscordBM has sane default behavior built around `retryPolicy`, as follows:
```

`DefaultDiscordClient` takes help from a `HTTPRateLimiter` actor to handle rate-limits.
`HTTPRateLimiter` keeps track of bucket-infos that are available in request headers. Before each request there is a call to `HTTPRateLimiter` that asks "can I do a request now?" and the rate-limiter, considering Discord docs' notes about how exactly Discord rate-limits work, decides to allow a request or not.   
For example if a bucket is exhausted, the `HTTPRateLimiter` will have no choice but to reject the request, Right?   
That's not exactly what happens in DiscordBM. If there is a request that is disallowed by the bucket info, `HTTPRateLimiter` instead tries to return the time amount the `DiscordClient` needs to wait before making the request. Then `DiscordClient` takes a look at the `retryPolicy`, and if the `retryPolicy` specifies that requests failed with the `429 Too Many Requests` header can be retried based on headers, `DiscordClient` will just wait the time, and make the request after! This is best of the both worlds:

> - Even if you make a lot of requests in a loop, `DiscordClient` not only won't fail, but also won't even let users notice anything. To users it will look like as if they don't even have a rate limit to worry about!

Reviewer: What if the number of messages per `<timeunit>` exceeds the amount that can be regained at the rate limiter?

Author: So to be clear, the whole mechanism doesn't stop you from exceeding your rate-limit.
It makes that happen way less often, but it can't stop you.

I didn't get what you exactly mean, but if you do requests in a loop, it should never fail. What happens is that it for example makes 5 requests immediately, then notices it needs to wait 4.5s before doing the next, and DiscordBM will just wait out that 4.5 before doing the next request, an then continues the loop.

Was that clear? Should i change what is in the proposal?

- On the other hand, `DiscordClient` didn't make a request that would have gotten a `429` back from Discord, so that's another win.

For all public API, DiscordBM comes with a lot of unit tests as well as integration tests.
This means:

- `BotGatewayManager` has [extensive tests](https://github.com/DiscordBM/DiscordBM/blob/main/Tests/IntegrationTests/GatwayConnection.swift) to make sure it connects and stays connected.
- Most of ~200 `DiscordClient` REST API functions [have a test](https://github.com/DiscordBM/DiscordBM/blob/main/Tests/IntegrationTests/DiscordClient.swift) that sends requests to Discord to make sure the functions do actually work in practice.

> If you want to more about the DiscordBM's public APIs, please refer to the GitHub [README](https://github.com/DiscordBM/DiscordBM#readme).

Reviewer: ```suggestion
If you want to learn more about the DiscordBM's public APIs, please refer to the GitHub [README](https://github.com/DiscordBM/DiscordBM#readme).
```
